### PR TITLE
[metrics] Improve the gauges API to allow closing a gauge [PLEN-107]

### PR DESCRIPTION
**Pull request by nicu-da**
This ensures gauges can be removed once they are no longer needed/used.

### CODE REVIEW

File: ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala [hunk 1]
```diff
@@ -53,7 +53,8 @@ class IndexerMetrics(prefix: MetricName, factory: MetricsFactory) {
                     |can be negative.""",
     qualification = Debug,
   )
-  val currentRecordTimeLag: Gauge[Long] = DropwizardGauge(prefix :+ "current_record_time_lag", null)
+  val currentRecordTimeLagForDocs: Gauge[Long] =
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala:56]: Just wondering, what prevents us from defining this at the place where it actually gets updated?

File: observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala [hunk 2]
```diff
@@ -165,15 +168,29 @@ object InMemoryMetricsFactory extends InMemoryMetricsFactory {
...
+      checkClosed()
+      value.get()
+    }
 
-    override def getValue: T = value.get()
+    override def close(): Unit = closed.set(true)
 
-    override def updateValue(f: T => T): Unit = discard(value.updateAndGet((t: T) => f(t)))
+    private def checkClosed(): Unit =
+      if (closed.get()) throw new IllegalArgumentException("Already closed")
```
Comment (miklos-da) [inline observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala:193]: Wouldn't an `IllegalStateException` be more appropriate here? I.e., it is inappropriate to call `getValue` or `updateValue` on a closed `Gauge`.

**Review by miklos-da — APPROVED** [review]

File: ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala (hunk 1 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala:56]: We didn't have support for the async gauge as of yet as it wasn't  returning anything. If we were to make `CloseableGauge` a metric handle we wouldn't require this anymore.

File: observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala (hunk 2 above)
Comment (nicu-da) [reply observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala:193]: Good point

File: ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala (hunk 1 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala:56]: Will handle that in a separate PR

**Pull request merged**